site/content/contribute/server/style-guide.md

```diff
@@ -0,0 +1,130 @@
+---
+title: "Go Style Guide"
+date: 2021-01-12T16:00:00+0530
+weight: 3
+subsection: Server
+---
+
+Go is a more opinionated language than many others when it comes to coding style. The compiler enforces some basic stylistic elements, such as the removal of unused variables and imports. Many others are enforced by the `gofmt` tool, such as usage of white-space, semicolons, indentation and alignment. The `gofmt` tool is run over all code in the Mattermost Server CI pipeline, and code which is not consistent with the formatting enforced by `gofmt` will not be accepted into the repository.
+
+Despite this, there are still many areas of coding style which are not dictated by these tools. Rather than reinventing the wheel, we are adopting [Effective Go](https://golang.org/doc/effective_go.html) as a basis for our style guide. On top of that, we also follow the guidelines laid out by the Go project at [CodeReviewComments](https://github.com/golang/go/wiki/CodeReviewComments).
+
+However, at present, some of the guidelines from these sources come into conflict with existing patterns that are present in our codebase which cannot immediately be corrected due to the need to maintain backwards compatibility.
+
+This document, which should be read in conjunction with [Effective Go](https://golang.org/doc/effective_go.html) and [CodeReviewComments](https://github.com/golang/go/wiki/CodeReviewComments), outlines the small number of exceptions we make to maintain backwards compatibility, as well as a number of additional stylistic rules we have adopted on top of those external recommendations.
+
+### Model package
+
+The `model` package is our public API which is consumed by many plugins and 3rd party integrations. The need to maintain backwards-compatibility for these external users of the `model` package prevents us from immediately bringing it into compliance with some of the rules outlined in this document.
+
+In order to avoid delaying the adoption of these rules in the wider codebase, we have chosen to temporarily exempt only the `model` package from certain rules (indicated below). However, outside of the `model` package, these rules should be followed in all new or modified code.
+
+### Application of Guidelines
+
+Aside from the specific exceptions made for backward compatibility in the `model` package, all new commits should follow the rules as outlined in this and the linked documents, for both new and modified code.
+
+This does not, however, mean that a developers is *required* to fix any surrounding code that contravenes the rules in the style guide. It is encouraged to keep fixing things as you go, but it is not compulsory to do so. Reviewers should refrain from asking for stylistic changes in surrounding code if the submitter has not included them in their pull request.
+
+## Guidelines
```

Comment:
Wondering if we should change this to `The Rules` instead? What do you think?

Reply:
I'd prefer to use `Guidelines`. `Rules` bring a more strict connotation. You can strictly apply the style rules, but the interfaces and concurrency stuff cannot be rules. They need to be considered as guidelines which may need to be skipped in cases.

Reply:
:+1: